torch/csrc/jit/tensorexpr/analysis.h

```diff
@@ -248,6 +248,194 @@ class ModifiesVarChecker : public IRVisitor {
   bool found_{false};
 };
 
+// This indicates the path from root to a stmt A in AST. Specifically, it
+// consists of a vector of integers: each integer represents the number of stmts
+// before A's ancestor or A in the block. For example, "1:1:0" points to stmt
+// "d[i] += e[i, j];" in the following code.
+//
+// c = a/b;
+// for i
+//  d[i] = 0;
+//  for j
+//    d[i] += e[i, j];
+class StmtIndex {
+ public:
+  void append(int32_t id) {
+    stmt_index_.emplace_back(id);
+  }
+  void pop() {
+    stmt_index_.pop_back();
+  }
+
+  const std::vector<int32_t>& getStmtIndex() {
+    return stmt_index_;
+  }
+
+  bool operator==(StmtIndex& index) {
+    auto compare = index.getStmtIndex();
+    if (stmt_index_.size() != compare.size()) {
+      return false;
+    }
+    int size = compare.size();
+    for (int i = 0; i < size; i++) {
+      if (stmt_index_.at(i) != compare.at(i)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  bool operator<(StmtIndex& index) {
+    auto compare = index.getStmtIndex();
+    int size = stmt_index_.size() < compare.size() ? stmt_index_.size()
+                                                   : compare.size();
+    for (int i = 0; i < size; i++) {
+      if (stmt_index_.at(i) > compare.at(i)) {
+        return false;
+      }
+    }
+    return !(*this == index);
+  }
+
+  bool operator>(StmtIndex& index) {
+    auto compare = index.getStmtIndex();
+    int size = stmt_index_.size() < compare.size() ? stmt_index_.size()
+                                                   : compare.size();
+    for (int i = 0; i < size; i++) {
+      if (stmt_index_.at(i) < compare.at(i)) {
+        return false;
+      }
+    }
+    return !(*this == index);
+  }
+
+  std::string getStmtIndexString() {
+    std::string cstr = "";
+    for (int i = 0; i < stmt_index_.size(); i++) {
+      cstr += std::to_string(stmt_index_.at(i));
+      cstr += ":";
+    }
+    return cstr;
+  }
+
+ private:
+  std::vector<int32_t> stmt_index_;
+};
+
+// Traverse the root stmt to identify the position of each stmt in the AST,
+// i.e., a StmtIndex
+class StmtIndexer : public IRVisitor {
+ public:
+  StmtIndex getStmtIndex() {
+    return stmt_index_;
+  }
+
+ private:
+  void visit(BlockPtr v) {
+    int count = 0;
+    for (StmtPtr s : *v) {
+      stmt_index_.append(count);
+      s->accept(this);
+      stmt_index_.pop();
+      count++;
```

Comment:
Why do we need `count`?

Reply:
We use `count` to count the number of stmts we have traversed in the block. Stmts in a block is organized as a list. We need a separate integer to track how many stmts have been traversed.